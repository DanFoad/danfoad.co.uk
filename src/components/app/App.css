/**** GENERAL ****/
* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}
*:before, *:after {
  box-sizing: inherit;
}
html, body, #root {
  width: 100%;
  height: 100%;
  font-size: 62.5%;
}

/**** APP ****/
.app {
  width: 100%;
  height: 100%;
  font-family: "Open Sans", sans-serif;
}