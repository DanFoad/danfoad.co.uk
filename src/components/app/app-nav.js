import React from 'react';

import { t } from '../../services/label-service';

export default class AppNav extends React.Component {
  convertIndex(index) {
    return index++ < 9 ? '0' + index : index;
  }

  getNavLinks() {
    const links = t('app.nav.links', true);
    return links.map((link, index) => {
      return (
        <li key={index}>
          <span className="index">{this.convertIndex(index)}</span>
          {link}
        </li>
      );
    });
  }

  getNavContents() {
    return (
      <nav className="main-nav">
        <h1 className="main-nav__title">
          <a href="/">
            {t('general.name')} :: <span>{t('general.role')}</span>
          </a>
          <div className="main-nav__edge" />
        </h1>
        <div className="main-nav__slant" />
        <div className="main-nav__links--container">
          <ul className="main-nav__links">{this.getNavLinks()}</ul>
        </div>
      </nav>
    );
  }

  render() {
    return this.getNavContents();
  }
}
