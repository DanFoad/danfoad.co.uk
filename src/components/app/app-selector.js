import { createSelector } from 'reselect';

export const appSelector = createSelector(
  [state => state.app],
  app => {
    return {
      labelsLoading: app.labelsLoading
    };
  }
);
