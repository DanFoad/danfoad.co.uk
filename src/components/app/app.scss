/**** GENERAL ****/
* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}
*:before,
*:after {
  box-sizing: inherit;
}
body {
  font-size: 62.5%;
}
html,
body,
#root {
  width: 100%;
  height: 100%;
}

/**** APP ****/
.app {
  width: 100%;
  height: 100%;
  font-family: 'Open Sans', sans-serif;
}

/**** NAV ****/
.main-nav {
  display: grid;
  grid-template-columns: [start] auto [header-start] 400px [header-slant] 48px [header-mid] 752px [header-end] auto [end];
  height: 48px;
  position: fixed;
  top: 16px;
  left: 0;
  right: 0;
  margin: 0 auto;
}
.main-nav__title {
  border-bottom: 1px solid #ffffff;
  padding-bottom: 16px;
  position: relative;
  padding-right: 8px;
  grid-column: start / header-slant;
}
.main-nav__title a {
  text-transform: uppercase;
  display: inline-block;
  background-color: rgba(255, 255, 255, 0.75);
  color: #1e282e;
  line-height: 32px;
  padding: 0 16px;
  height: 32px;
  width: 100%;
  text-align: right;
  text-decoration: none;
}
.main-nav__title span {
  color: rgba(0, 0, 0, 0.6);
  font-weight: normal;
}
.main-nav__edge {
  position: absolute;
  right: -24px;
  top: 0;
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 32px 32px 0 0;
  border-left-width: 100%;
  border-color: rgba(255, 255, 255, 0.75) transparent transparent transparent;
}
.main-nav__slant {
  grid-column: header-slant / header-mid;
  width: 48px;
  height: 1px;
  transform: rotate(-45deg) scaleX(1.4);
  transform-origin: bottom left;
  position: relative;
  background: #ffffff;
  top: 48px;
  right: -1px;
}
.main-nav__links--container {
  grid-column: header-mid / end;
  display: grid;
  grid-template-columns: [nav-start] 752px [nav-end] auto [end];
  border-top: 1px solid #ffffff;
}
.main-nav__links {
  grid-column: nav-start / nav-end;
  list-style: none;
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  justify-content: flex-end;
  flex: 1;
  font-family: 'Space Mono', monospace;
}
.main-nav__links li {
  padding: 0 6px;
  margin: 0 24px;
  font-size: 1.3em;
  cursor: pointer;
  position: relative;
}
.main-nav__links li:hover {
  color: #1e282e;
  background-color: #63859b;
}
.main-nav__links li:hover:before {
  content: '';
  width: 16px;
  height: 100%;
  background-color: rgb(120, 184, 226);
  position: absolute;
  left: -20px;
  top: 0;
  opacity: 0.8;
  animation: flash 0.5s ease-in-out infinite alternate;
}
@keyframes flash {
  from {
    opacity: 0;
  }
  to {
    opacity: 0.8;
  }
}
