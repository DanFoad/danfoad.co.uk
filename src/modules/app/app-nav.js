import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import { t } from '../../services/label-service';

export class AppNav extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      menuVisible: false,
      menuClass: '',
      links: {
        'app.nav.links.home': '/',
        'app.nav.links.projects': '/projects/',
        'app.nav.links.cv': '/cv/',
        'app.nav.links.blog': '/blog/',
        'app.nav.links.contact': '/contact-me/',
      },
    };
  }

  getNavLinks() {
    const links = Object.keys(this.state.links).map(l => t(l));
    return links.map((link, index) => {
      return (
        <Link to={Object.values(this.state.links)[index] || '#'} key={index}>
          <li>
            <span className="index">{this.convertIndex(index)}</span>
            {link}
          </li>
        </Link>
      );
    });
  }

  getNavContents() {
    const navClass = this.props.transparentBackground
      ? 'main-nav--transparent'
      : '';

    return (
      <nav className={`main-nav ${navClass}`}>
        <h1 className="main-nav__title">
          <a href="/">
            {t('general.name')}
            <span className="separator"> :: </span>
            <span>{t('general.role')}</span>
          </a>
          <div className="main-nav__edge" />
        </h1>
        <div className="main-nav__slant" />
        <div className="main-nav__links--desktop">
          <ul className="main-nav__links">{this.getNavLinks()}</ul>
        </div>
        <div className="main-nav__links--mobile">
          <div className="main-nav__links--menu" onClick={this.toggleMenu}>
            <img src="/img/menu.png" alt="" />
          </div>
          {this.state.menuVisible && (
            <div
              className="main-nav__links--overlay"
              onClick={this.toggleMenu}
            />
          )}
          <ul className={`main-nav__links ${this.state.menuClass}`}>
            {this.getNavLinks()}
          </ul>
        </div>
      </nav>
    );
  }

  toggleMenu = () => {
    this.setState(prevState => ({
      menuVisible: !prevState.menuVisible,
      menuClass: prevState.menuClass.length === 0 ? 'open' : '',
    }));
  };

  convertIndex(index) {
    return index < 9 ? `0${index + 1}` : index + 1;
  }

  render() {
    return this.getNavContents();
  }
}

AppNav.propTypes = {
  transparentBackground: PropTypes.bool,
};
