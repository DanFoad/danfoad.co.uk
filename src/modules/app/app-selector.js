import { createSelector } from 'reselect';

export const appSelector = createSelector(
  [state => state.app],
  app => {
    return {
      labelsLoading: app.labelsLoading,
      navExpanded: app.navExpanded,
      navLinks: app.navLinks,
      transparentNav: app.transparentNav,
    };
  }
);
