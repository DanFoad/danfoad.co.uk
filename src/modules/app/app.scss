@import '../../index';

/**** GENERAL ****/
* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}
*:before,
*:after {
  box-sizing: inherit;
}
body {
  font-size: 62.5%;
  font-family: 'Open Sans', sans-serif;
  background-color: $dark1;
}
html,
body,
#root {
  width: 100%;
  height: 100%;
}

/**** NAV ****/
.main-nav {
  display: grid;
  grid-template-columns: [start] auto [header-start] 400px [header-slant] 48px [header-mid] 752px [header-end] auto [end];
  height: 64px;
  position: fixed;
  padding-top: 16px;
  top: 0;
  left: 0;
  right: 0;
  margin: 0 auto;
  z-index: 10;
  background: rgba($dark1, 0.9);

  &--transparent {
    background: transparent;
  }
}
.main-nav__title {
  border-bottom: 1px solid $light3;
  padding-bottom: 16px;
  position: relative;
  padding-right: 8px;
  grid-column: start / header-slant;
  font-family: 'Space Mono', sans-serif;
}
.main-nav__title a {
  text-transform: uppercase;
  display: inline-block;
  background-color: $light3;
  color: $dark1;
  line-height: 32px;
  padding: 0 16px;
  height: 32px;
  width: 100%;
  text-align: right;
  text-decoration: none;
}
.main-nav__title span {
  color: rgba(0, 0, 0, 0.6);
  font-weight: normal;
}
.main-nav__title .separator {
  color: $dark1;
  position: relative;
  top: -2px;
}
.main-nav__edge {
  position: absolute;
  right: -24px;
  top: 0;
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 32px 32px 0 0;
  border-left-width: 100%;
  border-color: $light3 transparent transparent transparent;
}
.main-nav__slant {
  grid-column: header-slant / header-mid;
  width: 48px;
  height: 1px;
  transform: rotate(-45deg) scaleX(1.4);
  transform-origin: bottom left;
  position: relative;
  background: $light3;
  top: 48px;
  right: -1px;
}
.main-nav__links--desktop {
  grid-column: header-mid / end;
  display: grid;
  grid-template-columns: [nav-start] 752px [nav-end] auto [end];
  border-top: 1px solid $light3;
}
.main-nav__links {
  grid-column: nav-start / nav-end;
  list-style: none;
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  justify-content: flex-end;
  flex: 1;
  font-family: 'Space Mono', monospace;
}
.main-nav__links a {
  display: inline-flex;
  height: 100%;
  align-items: center;
  padding: 0 24px;
  color: $light3;
  text-decoration: none;
}
.main-nav__links li {
  padding: 0 6px;
  font-size: 1.3em;
  text-transform: uppercase;
  cursor: pointer;
  position: relative;
  display: inline-flex;
  align-items: center;
  transition: color 0.2s;
}
.main-nav__links a:hover li {
  color: $dark1;
}
.main-nav__links li:before {
  content: '';
  width: 100%;
  height: 0;
  background-color: $yellow;
  position: absolute;
  left: 0;
  bottom: 0;
  z-index: -1; 
  transition: height 0.2s; 
}
.main-nav__links a:hover li:before {
  height: 100%;
}
.main-nav__links .index {
  color: $yellow;
  margin-right: 12px;
  font-size: 0.8em;
  transition: color 0.2s;
}
.main-nav__links a:hover .index {
  color: $dark4;
}
.main-nav__links--mobile {
  grid-column: header-end / end;
  display: none;
  border-top: 1px solid $light3;

  .main-nav__links--menu {
    height: 100%;
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    cursor: pointer;
    padding-right: 8px;
    -webkit-appearance: none;
    -webkit-tap-highlight-color: transparent;
  }

  .main-nav__links--menu img {
    height: 24px;
    width: 24px;
  }

  .main-nav__links {
    flex-flow: column nowrap;
    align-items: flex-end;
    justify-content: flex-start;
    position: absolute;
    top: 72px;
    right: -192px;
    z-index: 10;
    transition: all 0.2s;
  }

  .main-nav__links.open {
    right: 0;
    transition: all 0.2s;
  }

  .main-nav__links--overlay {
    position: fixed;
    left: 0;
    top: 0;
    right: 0;
    bottom: 0;
    width: 100%;
    height: 100%;
    margin: auto;
    z-index: 9;
    background: rgba(0, 0, 0, 0.6);
  }

  .main-nav__links li {
    width: 144px;
    position: relative;
    display: flex;
    justify-content: flex-end;
    padding: 8px 0;
    color: $dark1;
  }

  .main-nav__links .index {
    position: absolute;
    left: -16px;
    top: 0;
    bottom: 0;
    margin: auto;
    display: inline-flex;
    align-items: center;
    color: $dark4;
  }

  .main-nav__links li:before {
    content: '';
    width: 8px;
    height: 100%;
    background-color: $yellow;
    position: absolute;
    right: -24px;
    top: 0;
    left: auto;
    bottom: 0;
    z-index: -1; 
    transition: width 0.2s; 
  }

  .main-nav__links a:hover li:before {
    width: calc(100% + 48px);
  }

  .main-nav__links li:after {
    content: '';
    width: calc(100% + 48px);
    height: 100%;
    background-color: $light3;
    position: absolute;
    right: -24px;
    top: 0;
    left: auto;
    bottom: 0;
    z-index: -2;
  }
}

@media all and (max-width: 1200px) {
  .main-nav {
    grid-template-columns: [start] auto [header-slant] 48px [header-end] 96px [end];
  }
  .main-nav__links--desktop {
    display: none;
  }
  .main-nav__links--mobile {
    display: block;
  }
  .main-nav__slant {
    grid-column: header-slant / header-end;
  }
  .main-nav__title a {
    text-align: left;
  }
  .home-header__contents {
    align-items: center;
  }
}

@media all and (max-width: 600px) {
  .main-nav__title span {
    display: none;
  }
}

/**** SECTIONS ****/
.section {
  max-width: 1200px;
  width: 100%;
  align-self: center;
  margin: 64px auto;
}