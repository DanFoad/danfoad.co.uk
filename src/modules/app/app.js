import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Switch, Route } from 'react-router-dom';
import PropTypes from 'prop-types';

import { init as initLabels } from '../../services/label-service';
import { setTransparentNav } from '../../reducers/app-reducers';
import { Home } from '../home/home-component';
import { CV } from '../cv/cv-component';

import { AppNav } from './app-nav';
import { appSelector } from './app-selector';

import './app.scss';
import '../../components/headers/headers.scss';

class AppComponent extends Component {
  componentDidMount() {
    this.props.initLabels();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.location.pathname === this.props.location.pathname) return;

    if (this.props.transparentNav) this.props.setTransparentNav(false);
  }

  render() {
    if (this.props.labelsLoading) return <div />;

    return (
      <React.Fragment>
        <AppNav
          navLinks={this.props.navLinks}
          expanded={this.props.navExpanded}
          transparentBackground={this.props.transparentNav}
        />
        <Switch>
          <Route exact path="/" component={Home} />
          <Route exact path="/cv/" component={CV} />
        </Switch>
      </React.Fragment>
    );
  }
}

AppComponent.propTypes = {
  labelsLoading: PropTypes.bool,
  initLabels: PropTypes.func.isRequired,
  navExpanded: PropTypes.bool.isRequired,
  navLinks: PropTypes.objectOf(PropTypes.string).isRequired,
  transparentNav: PropTypes.bool,
  setTransparentNav: PropTypes.func.isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired,
    search: PropTypes.string.isRequired,
    hash: PropTypes.string.isRequired,
    state: PropTypes.object,
  }).isRequired,
};

export const App = connect(
  appSelector,
  {
    initLabels,
    setTransparentNav,
  }
)(AppComponent);
