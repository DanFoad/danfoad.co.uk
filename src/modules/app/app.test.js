import React from 'react';
import { shallow } from 'enzyme';

import { App as WrappedApp } from './app';

const App = WrappedApp.WrappedComponent;

jest.mock('react-router-dom');

describe('app-container', () => {
  let props;

  beforeEach(() => {
    props = {
      location: {
        pathname: '/',
        search: '',
        hash: '',
      },
      labelsLoading: false,
      navExpanded: false,
      navLinks: {
        'app.nav.links.home': '/',
      },
      initLabels: jest.fn(),
      setTransparentNav: jest.fn(),
    };
  });

  it('should initialise labels on mount', () => {
    shallow(<App {...props} />);
    expect(props.initLabels.mock.calls.length).toBe(1);
  });

  it('should set the nav back to non-transparent on page change', () => {
    const wrapper = shallow(<App {...props} transparentNav />);
    wrapper.setProps({ location: { ...props.location, pathname: '/test' } });

    expect(props.setTransparentNav).toBeCalledWith(false);
  });

  it('should render an empty div if labels still loading', () => {
    expect(shallow(<App {...props} labelsLoading />)).toMatchSnapshot();
  });

  it('should render the app once labels are ready', () => {
    expect(shallow(<App {...props} />)).toMatchSnapshot();
  });
});
