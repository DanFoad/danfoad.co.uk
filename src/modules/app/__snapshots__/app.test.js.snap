// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app-container should render an empty div if labels still loading 1`] = `<div />`;

exports[`app-container should render the app once labels are ready 1`] = `
<Fragment>
  <AppNav />
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/cv/"
    />
  </Switch>
</Fragment>
`;
